### ISSUE TITLE
`Phoenix.Controller.send_download/3` incorrectly escapes spaces in filenames

### ISSUE THREAD

**Issue by bencates** [phoenixframework/phoenix#5478]
### Environment

* Elixir version (elixir -v): Erlang/OTP 25, Elixir 1.14.4
* Phoenix version (mix deps): 1.7.2

### Expected behavior

When hitting a controller route that looks like this:

```elixir
def download(conn, _params) do
  send_download(conn, {:binary, "test"}, filename: "name with spaces.txt")
end
```

I would expect the user to be prompted to download a file named `name with spaces.txt`.

### Actual behavior

The user is prompted to download a file named `name+with+spaces.txt`

### Proposed solution

Using RFC-3986 encoding instead of "x-www-form-urlencoded"; i.e. encoding the space character as `%20` instead of `+`. The following does preserve spaces:

```elixir
def download(conn, _params) do
  send_download(conn, {:binary, "test"},
    filename: URI.encode("name with spaces.txt", %URI.char_unreserved?/1),
    encode: false
  )
end
```

I don't believe using `URI.char_unreserved?/1` instead of the default `URI.char_unescaped?/1` is actually necessary for security reasons but it is the smallest breaking change from `URI.encode_www_form/1`.

### Previous discussions

Discussed in #3344 where the `encode: false` option was introduced. The option to use `URI.encode/2` was I believe dismissed in error.

I suspect when testing whether `URI.encode/2` would work the filename accidentally got encoded twice (i.e. `filename="some%2520file.txt"`), which is what happens if you manually `URI.encode` the filename without also disabling the automatic encoding. The browser will decode that - once - into `some%20file.txt`.

**Comment by mcrumm** [comment]
The filename parameter of Content-Disposition only supports ASCII characters without introducing quirks across different user agents, so the current encoding is technically more correct, though perhaps it is not the most desirable default.

I think we may want to revisit send_download in the context of RFC 6266, specifically the guidance in [[RFC 6266], Appendix D](https://www.rfc-editor.org/rfc/rfc6266#appendix-D), and include a "filename*" parameter that uses an encoding defined in [RFC 5987](https://www.rfc-editor.org/rfc/rfc5987).

tl;dr I think we want to produce a header in this shape:

```
Content-Disposition: attachment; filename*=utf-8''name%20with%20spaces.txt; filename="name+with+spaces.txt"

**Comment by bencates** [comment]
Yeah, adding a `filename*` param would be the ideal solution here. I'd be happy to put together a PR to do that with or without changes to how the existing `filename` param is treated.

I'm still not sure `filename="name+with+spaces.txt"` is reasonable, though. If I'm reading the RFC correctly, when the value of the `filename` attribute is quoted (which Phoenix always does) it can contain any (printable?) ASCII characters except a literal `"` (and maybe `\` and `%` since some UAs will treat those as escape sequences even though the spec doesn't say to). That might not produce a valid filename on whatever filesystem the client is using, but the spec is very clear that that's the client's problem and not the server's.

I think what we want is `filename="name with spaces.txt"`. If we are going to encode the value then I think there's good reason to prefer `name%20with%spaces.txt` to `name+with+spaces.txt`; some user agents (including both Chrome and Firefox) will demangle `%20` but as far as I know all of them will leave literal `+` characters in.

**Comment by mcrumm** [comment]
I first tested this out by _just_ switching to `URI.encode/1` and Safari fails to decode the filename properly, though Chrome and Firefox were okay. So we need to add filename* I'm just working thru some additional test cases.
